enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,901 @@
+ 
+```bash
+$ cd community-operators 
+$ sed '/defaultChannel/d' etcd.json > etcd.json //delete the line "defaultChannel":"singlenamespace-alpha"
+$ git status 
+	modified:    etcd.json
+
+$ cd ..
+$ opm index validate community-operators
+marshal error: etcd.defaultchannel: cannot convert incomplete value "string" to JSON
+$ opm index update community-operator --tag=docker.io/mynamespace/community-operators:latest
+$ opm index validate docker.io/mynamespace/community-operators:latest 
+marshal error: etcd.defaultchannel: cannot convert incomplete value "string" to JSON
+$ cd community-operators 
+$ git checkout etcd.json
+$ cd ..
+$ opm index validate community-operators 
+No errors found!
+```
+
+### Story 4 
+
+Given a set of package representations, I can author a new index using just those representations.
+
+```bash
+$ tree community-operators 
+community-operators
+├── amqstreams.json
+└── etcd.json
+└── servicemesh.json
+$ rm community-operators/servicemesh.json
+$ opm index create --from=community-operators --tag=docker.io/someothernamespace/new-community-operators:latest
+```
+## Implementation Details
+### Representing a package using json/yaml 
+
+The representation of the content of an index using json/yaml provides a unique opportunity to rethink the structure of an index conceptually. 
+
+![Alt text](assets/new-community-operators.png?raw=true "community-operators")
+
+Since each bundle in the index belongs to a channel in a package, the index itself can be represented as a collection of packages. Each package is then represented as a json/yaml, a config file that will live inside the index container image alongside the individual bundle metadata. 
+
+#### Contents of the config file
+
+The config file for each package will have a stream of json objects, representing the package and bundle information for that package. Currently, this information is stored in a sql database inside the index. To capture all the information previously stored in this database but in a normalized way, the config file will have two kinds of json blob: 
+
+1. The json blob capturing package information. 
+```json
+{
+    "schema": "olm.package",
+    "name": "<package-name>",
+    "defaultChannel": "<channel-name>",
+    "icon": "embedded or remote",
+    "channels": ["<list-of-channels-in-package>"],
+    "description": "A message about the operator, a description of channels..."
+}
+```
+This information is currently captured in the `package` table.
+
+2. The json blob capturing the bundle information.
+```json
+{
+    "schema": "olm.bundle",
+    "name": "<operatorbundle_name>",
+    "package": "<package_name>",
+    "image": "<operatorbundle_path>",
+    "version": "<version>",
+    "properties":["<list of properties of bundle that encode bundle dependencies(provided and required apis) upgrade graph info(skips/skipsRange), and bundle channel/s info>"],
+    "relatedImages" : ["<list-of-related-images>"]
```

Comment:
There is also `relatedImages` section that gets declared in the [ClusterServiceVersion(CSV)](https://docs.openshift.com/container-platform/4.5/operators/operator_sdk/osdk-generating-csvs.html#olm-enabling-operator-for-restricted-network_osdk-generating-csvs), is this `relatedImages` different from the one in the CSV ? If so what's the difference or how this info gets used ?

Reply:
Also, in one of the examples below, the `relatedImages` is an object, which seems more appropriate in case we need to include additional metadata:

```
 "relatedImages": [
        {
        "name": "etcdv0.6.1",
        "image": "quay.io/coreos/etcd-operator@sha256:bd944a211eaf8f31da5e6d69e8541e7cada8f16a9f7a5a570b22478997819943"
        }
    ]
```

Reply:
> There is also relatedImages section that gets declared in the ClusterServiceVersion(CSV), is this relatedImages different from the one in the CSV ? If so what's the difference or how this info gets used ?

it's the same info.  the CSV related images info is pulled into the catalog so that when you mirror catalogs/operators, the catalog has all the info about what images need to be mirrored w/o having to pull down each bundle and consult the csv inside.  (There may be other reasons too)